src/main/java/com/puppycrawl/tools/checkstyle/site/SiteUtil.java

```diff
@@ -0,0 +1,171 @@
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+///////////////////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle.site;
+
+import java.lang.reflect.Field;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Locale;
+import java.util.Set;
+
+import org.apache.maven.doxia.macro.MacroExecutionException;
+
+import com.puppycrawl.tools.checkstyle.ModuleFactory;
+import com.puppycrawl.tools.checkstyle.PackageNamesLoader;
+import com.puppycrawl.tools.checkstyle.PackageObjectFactory;
+import com.puppycrawl.tools.checkstyle.api.CheckstyleException;
+import com.puppycrawl.tools.checkstyle.checks.regexp.RegexpMultilineCheck;
+import com.puppycrawl.tools.checkstyle.checks.regexp.RegexpSinglelineCheck;
+import com.puppycrawl.tools.checkstyle.checks.regexp.RegexpSinglelineJavaCheck;
+
+/**
+ * Utility class for site generation.
+ */
+public final class SiteUtil {
+
+    /**
+     * Private utility constructor.
+     */
+    private SiteUtil() {
+    }
+
+    /**
+     * Gets the check's messages keys.
+     *
+     * @param module class to examine.
+     * @param checkInstance instance of the check.
+     * @return a list of checkstyle's module message fields.
+     * @throws MacroExecutionException if the attempt to read a protected class fails.
+     */
+    public static List<String> getCheckMessageKeys(Class<?> module, Object checkInstance)
+            throws MacroExecutionException {
+        try {
+            final List<String> checkstyleMessageKeys = new ArrayList<>();
+
+            // get all fields from current class
+            final Field[] fields = module.getDeclaredFields();
+
+            for (Field field : fields) {
+                if (field.getName().startsWith("MSG_")) {
+                    checkstyleMessageKeys.add(getFieldValue(field, checkInstance));
+                }
+            }
+
+            // deep scan class through hierarchy
+            final Class<?> superModule = module.getSuperclass();
+
+            if (superModule != null) {
+                checkstyleMessageKeys.addAll(getCheckMessageKeys(superModule, checkInstance));
+            }
+
+            // special cases that require additional classes
+            if (module == RegexpMultilineCheck.class) {
+                checkstyleMessageKeys.addAll(getCheckMessageKeys(
+                        Class.forName(
+                                "com.puppycrawl.tools.checkstyle.checks.regexp.MultilineDetector"
+                        ), checkInstance)
+                );
+            }
+            else if (module == RegexpSinglelineCheck.class
+                    || module == RegexpSinglelineJavaCheck.class) {
+                checkstyleMessageKeys.addAll(getCheckMessageKeys(
+                        Class.forName(
+                                "com.puppycrawl.tools.checkstyle.checks.regexp.SinglelineDetector"
```

Comment:
https://www.tutorialspoint.com/get-the-fully-qualified-name-of-a-class-in-java

Use java method to get fully qualified name.

Reply:
Same as https://github.com/checkstyle/checkstyle/pull/13423#discussion_r1272913610. Class is not public and I have no access to it
> 'com.puppycrawl.tools.checkstyle.checks.regexp.MultilineDetector' is not public in 'com.puppycrawl.tools.checkstyle.checks.regexp'. Cannot be accessed from outside package